### PR TITLE
3214 - Files and links per Cycle: Basic validation flow (RFC)

### PR DESCRIPTION
**Pull request by sorja**
Added:
- [x] Validation error message
- [x] Delete file button (unstyled)

https://github.com/openforis/fra-platform/assets/5508251/4e58624d-97e5-4221-8eb3-5d166d4ff2fb

### CODE REVIEW

File: src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx [hunk 1]
```diff
@@ -27,7 +34,7 @@ const Actions: React.FC = () => {
   return (
     <div className="repository-form__actions">
```
Comment (minotogna) [inline src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx:33]: the css class should be defined in its own css file

File: src/client/pages/CountryHome/Repository/Panel/InputFieldFile/InputFieldFile.tsx [hunk 2]
```diff
@@ -1,20 +1,30 @@
...
   const id = `repository_form-input-file`
+  const label = `common.file`
 
   return (
     <div className="repository-form__input-field">
       <label htmlFor={id} className="repository-form__label">
         {t(label)}
       </label>
       <FileUpload id={id} />
+      {repositoryItem?.fileUuid && (
```
Comment (minotogna) [inline src/client/pages/CountryHome/Repository/Panel/InputFieldFile/InputFieldFile.tsx]: should this belong to FileUpload ?

File: src/client/pages/CountryHome/Repository/Panel/InputFieldFile/InputFieldFile.tsx [hunk 3]
```diff
@@ -1,20 +1,30 @@
...
+        <button type="button" onClick={onRemoveFile}>
+          remove
```
Comment (minotogna) [inline src/client/pages/CountryHome/Repository/Panel/InputFieldFile/InputFieldFile.tsx]: and let's use ButtonDelete. what do you think ?

File: src/client/pages/CountryHome/Repository/Errors/Errors.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import { RepositoryItems } from 'meta/cycleData/repository'
+
+import { useRepositoryItem } from 'client/store/ui/repository'
+
+const Errors: React.FC = () => {
+  const { t } = useTranslation()
+  const repositoryItem = useRepositoryItem()
+  const validation = RepositoryItems.validate(repositoryItem)
+
+  const errors = Array.from(new Set(Object.values(validation || {})))
```
Comment (minotogna) [inline src/client/pages/CountryHome/Repository/Errors/Errors.tsx]: ideally we should highlight the cell with red border and tooltip message per field.
So maybe add the validation result to the store .
The validation gets updated automatically in slice, when a field or prop changes.
What do you think?

File: src/i18n/resources/en/generalValidation.js [hunk 5]
```diff
@@ -19,6 +19,9 @@ module.exports = {
   notEmpty: `Value cannot be empty`,
   otherLandExceedsExtentOfForest: `Exceeds Other land area (1a)`,
   remainingLandExceedsExtentOfForest: `Exceeds Remaining land area (1a)`,
+  repositoryValidationEitherFileOrLink: 'Only one of file or link is required',
```
Comment (minotogna) [inline src/i18n/resources/en/generalValidation.js]: Only one between file and link must be specified (this error highlights both, file and link in form)

File: src/i18n/resources/en/generalValidation.js [hunk 6]
```diff
@@ -19,6 +19,9 @@ module.exports = {
...
+  repositoryValidationFileOrLink: 'File or link is required',
```
Comment (minotogna) [inline src/i18n/resources/en/generalValidation.js]: One between link and file is required (highlights both fields as well)

File: src/meta/cycleData/repository/repositoryItems.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  if (fileUuid && link) {
+    return {
+      file: 'generalValidation.repositoryValidationEitherFileOrLink',
+      link: 'generalValidation.repositoryValidationEitherFileOrLink',
+    }
+  }
+
+  return undefined
+}
+export const RepositoryItems = {
```
Comment (minotogna) [inline src/meta/cycleData/repository/repositoryItems.ts]: let's call this RepositoryItemValidator ?

File: src/i18n/resources/en/generalValidation.js [hunk 8]
```diff
@@ -19,6 +19,9 @@ module.exports = {
...
+  repositoryValidationName: 'Name is required',
```
Comment (minotogna) [inline src/i18n/resources/en/generalValidation.js]: also, should we start refactoring this? let's have a new resource file validation.
`validation:{repositoryItem:{....}}`
What do you think? (We'll address all the messages in general validation later on)

File: src/client/pages/CountryHome/Repository/Errors/Errors.tsx (hunk 4 above)
Comment (sorja) [reply src/client/pages/CountryHome/Repository/Errors/Errors.tsx]: Sounds good...

the general implementation will be 

`state.validation = RepositoryItemValidator.validate(state.repositoryItem)`

File: src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx (hunk 1 above)
Comment (sorja) [reply src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx:33]: Not sure I follow?

Is it not?

image

File: src/client/pages/CountryHome/Repository/Panel/InputFieldFile/InputFieldFile.tsx (hunk 2 above)
Comment (sorja) [reply src/client/pages/CountryHome/Repository/Panel/InputFieldFile/InputFieldFile.tsx]: I will revisit a bit the fileuploader next PR if this is ok
This might be a bit misleading, but the functionality of the button is not related the UploadedFiles or `public.files` but to `RepositoryItem`

The functionality removes the `repositoryItem.fileUuid`

Case 1:
Delete button is shown:
We open the list of repository items in the platform
We click on pen icon
We click the delete icon
We see an error, either a link or a file is required

Case 2:
We add a new repository item
We upload a file
We see the delete icon
We add a link
We see an error, only one is required file vs link
We click the delete icon


in short: it deletes the fileUuid reference

Comment (sorja) [reply src/client/pages/CountryHome/Repository/Panel/InputFieldFile/InputFieldFile.tsx]: I am thinking of introducing `FileUpload.onDeleteFile`  with `onDeleteFile(uuid: string)` prop but I am not 100% sure on use case yet

**Comment by sorja** [comment]
@minotogna video updated

File: src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx [hunk 9]
```diff
@@ -18,7 +21,11 @@ const Actions: React.FC = () => {
   const upsertRepositoryItem = useUpsertRepositoryItem()
 
   const closePanel = useClosePanel()
-  const disabled = useIsRepositoryLoading()
+  const fileUploadLoading = useIsFileUploadLoading()
+  const valid = Objects.isEmpty(RepositoryItemValidator.validate(repositoryItem))
+  const isRepositoryLoading = useIsRepositoryLoading()
+  const disabled = isRepositoryLoading || fileUploadLoading
+  const disabledDone = disabled || !valid
```
Comment (minotogna) [inline src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx:26]: What's the difference between disabled and disabled done ?

File: src/client/pages/CountryHome/Repository/Panel/InputField/InputField.tsx [hunk 10]
```diff
@@ -2,6 +2,10 @@ import './InputField.scss'
 import React, { useCallback } from 'react'
 import { useTranslation } from 'react-i18next'
 
+import classNames from 'classnames'
+
+import { useRepositoryItemPropValidation } from 'client/store/ui/repository/hooks'
```
Comment (minotogna) [inline src/client/pages/CountryHome/Repository/Panel/InputField/InputField.tsx]: Can u please verify all the imports ?

File: src/client/pages/CountryHome/Repository/Panel/InputField/InputField.tsx [hunk 11]
```diff
@@ -33,6 +38,7 @@ const InputField: React.FC<Props> = (props: Props) => {
         name={name}
         type="text"
       />
+      <div className="repository-form__error">{error ? t(error) : ''}</div>
```
Comment (minotogna) [inline src/client/pages/CountryHome/Repository/Panel/InputField/InputField.tsx]: Maybe postfix error-label ?

File: src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx (hunk 9 above)
Comment (sorja) [reply src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx:26]: you cant click 'done' if there are errors in the form

File: src/meta/cycleData/repository/repositoryItemValidator .ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+export type RepositoryItemValidation =
+  | {
+      name?: string
+      fileUuid?: string
+      link?: string
+    }
+  | undefined
+
+const validate = (repositoryItem: Partial<RepositoryItem>): RepositoryItemValidation => {
```
Comment (minotogna) [inline src/meta/cycleData/repository/repositoryItemValidator .ts]: this can return RepositoryItemValidation | undefined. the type itself should not be undefined in this case.

File: src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx [hunk 13]
```diff
@@ -18,7 +19,12 @@ const Actions: React.FC = () => {
   const upsertRepositoryItem = useUpsertRepositoryItem()
 
   const closePanel = useClosePanel()
-  const disabled = useIsRepositoryLoading()
+  const fileUploadLoading = useIsFileUploadLoading()
+  const repositoryItemValidation = useRepositoryItemValidation()
+  const valid = Objects.isEmpty(repositoryItemValidation)
```
Comment (minotogna) [inline src/client/pages/CountryHome/Repository/Panel/Actions/Actions.tsx]: we can export a hook useIsRepositoryItemValid directly. what do you think?

**Review by minotogna — APPROVED** [review]
